### PR TITLE
enabling bam index caching in memory for GATKTool

### PR DESCRIPTION
**Pull request by lbergelson**
this seems to resolve the extremely slow runtimes when using large interval lists #2356

### CODE REVIEW

**Comment by lbergelson** [comment]
@LeeTL1220 4.alpha.2-146-gd9046b4-SNAPSHOT is uploading now.  Give it a shot.

**Comment by LeeTL1220** [comment]
@lbergelson Trying it now...

**Comment by droazen** [comment]
If we go with this solution we'll need to carefully study the effect on memory usage -- BQSR in particular is run in a lower-memory environment than previously.

**Comment by lbergelson** [comment]
@droazen I thought you might say that.  We might be able to come up with an intermediate caching solution where not the entire index is cached.  Although the index isn't really that big, so I'm not sure if it's a big deal.

**Comment by LeeTL1220** [comment]
@lbergelson @droazen How about a flag to turn it off and on?

**Comment by droazen** [comment]
Let's just measure the effect on memory usage with a single bam. Can use something like `PrintReads` to do the test. Then we can make a decision based on the result.

**Comment by lbergelson** [comment]
@droazen I'm not sure what sort of test I can have for this change.  Any thoughts? 

It's good that I looked at this again, it had a stupid bug in it the way I did it at first.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 1]
```diff
@@ -203,13 +204,18 @@ void initializeReference() {
      */
     void initializeReads() {
         if (! readArguments.getReadFiles().isEmpty()) {
-            SamReaderFactory factory = SamReaderFactory.makeDefault().validationStringency(readArguments.getReadValidationStringency());
+            final SamReaderFactory factory = SamReaderFactory.makeDefault().validationStringency(readArguments.getReadValidationStringency());
             if (hasReference()) { // pass in reference if available, because CRAM files need it
-                factory = factory.referenceSequence(referenceArguments.getReferenceFile());
+                factory.referenceSequence(referenceArguments.getReferenceFile());
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Recommend reverting to `factory = factory.referenceSequence()` out of paranoia that the return value might not be `this`

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 2]
```diff
@@ -96,6 +91,12 @@
     @Argument(fullName = StandardArgumentDefinitions.CLOUD_INDEX_PREFETCH_BUFFER_LONG_NAME, shortName = StandardArgumentDefinitions.CLOUD_INDEX_PREFETCH_BUFFER_SHORT_NAME, doc = "Size of the cloud-only prefetch buffer (in MB; 0 to disable). Defaults to cloudPrefetchBuffer if unset.", optional=true)
     private int cloudIndexPrefetchBuffer = -1;
 
+    @Argument(fullName = StandardArgumentDefinitions.DISABLE_BAM_INDEX_CACHING_LONG_NAME,
+            shortName = StandardArgumentDefinitions.DISABLE_BAM_INDEX_CACHING_SHORT_NAME,
+            doc = "If true, don't cache bam indexes, this will save memory but may decrease performance if many intervals are specified.",
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Add note that this argument has no effect if intervals are not specified.

**Comment by droazen** [comment]
Merge after addressing final comment :+1:

**Pull request merged**